docs/src/main/sphinx/develop/connectors.rst

```diff
@@ -223,10 +223,79 @@ features, like:
 Note that data modification also requires implementing
 a :ref:`connector-page-sink-provider`.
 
+When Trino receives a ``SELECT`` query, it parses it into an Intermediate
+Representation (IR). Then, during optimization, it checks if connectors
+can handle operations related to SQL clauses by calling one of the following
+methods of the ``ConnectorMetadata`` service:
+
+* ``applyLimit``
+* ``applyTopN``
+* ``applyFilter``
+* ``applyProjection``
+* ``applySample``
+* ``applyAggregation``
+* ``applyJoin``
+* ``applyTableFunction``
+* ``applyTableScanRedirect``
+
+Connectors can indicate that they don't support a particular pushdown or that
+the action had no effect by returning ``Optional.empty()``. Connectors should
+expect these methods to be called multiple times during the optimization of
+a given query.
+
+.. warning::
+
+  It's critical for connectors to return ``Optional.empty()`` if calling
+  this method has no effect for that invocation, even if the connector generally
+  supports a particular pushdown. Doing otherwise can cause the optimizer
+  to loop indefinitely.
+
+Otherwise, these methods return a result object containing the connectors
+table handle. The connector can store any kind of information in the table
+handle, as it's passed to other services it implements, like
+the ``ConnectorRecordSetProvider`` or ``ConnectorPageSourceProvider``.
+
+Limit and top-N pushdown
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+When executing a ``SELECT`` query with ``LIMIT`` or ``ORDER BY`` clauses,
+the query plan may contain a ``Sort`` or ``Limit`` operations.
+
+When the plan contains a ``Sort`` operation, the engine
+tries to push down the limit into the connector by calling the ``applyTopN``
+method of the connector metadata service. If there's no ``Sort`` operation, but
+only a ``Limit``, the ``applyLimit`` method is called, and the connector can
+return results in an arbitrary order.
+
+If the connector could benefit from the information passed to these methods but
+can't guarantee that it's be able to produce fewer rows than the provided
+limit, it should return a non-empty result containing a new handle for the
+derived table and the ``limitGuaranteed`` (in ``LimitApplicationResult``) or
+``topNGuaranteed`` (in ``TopNApplicationResult``) flag set to false.
+
+If the connector can guarantee to produce fewer rows than the provided
+limit, it should return a non-empty result with the "limit guaranteed" or
+"topN guaranteed" flag set to true.
+
+.. note::
+
+  The ``applyTopN`` is the only method that receives sort items from the
+  ``Sort`` operation. It's called even if there's no ``Limit`` operation.
```

Comment:
I'm not sure I understand what this is trying to convey.

Reply:
I removed the last sentence. I found it surprising that there's no `applySort` and I did not expect `applyTopN` to be the only way of pushing down ordering. Do you think it's worth keeping this note?